test/companion_web/controllers/auth_controller_test.exs

```diff
@@ -0,0 +1,76 @@
+defmodule CompanionWeb.AuthControllerTest do
+  use CompanionWeb.ConnCase
+
+  @ueberauth_auth %{
+    credentials: %{token: "aifoejwaofjaowj34"},
+    info: %{email: "test@example.org", urls: %{website: "example.org"}},
+    provider: :google
+  }
+
+  test "redirects user to Google for authentication", %{conn: conn} do
+    conn = get(conn, "/auth/google")
+    assert redirected_to(conn, 302)
+  end
+
+  test "creates user from Google information", %{conn: conn} do
+    conn =
+      conn
+      |> assign(:ueberauth_auth, @ueberauth_auth)
+      |> get("/auth/google/callback")
+
+    assert get_flash(conn, :info) == "Successfully signed in!"
+
+    assert get_session(conn, :user) == %{
+             email: @ueberauth_auth.info.email,
+             provider: :google
+           }
+
+    assert redirected_to(conn) == "/"
+  end
+
+  test "doesn't log use in if Google auth fails", %{conn: conn} do
+    conn =
+      conn
+      |> assign(:ueberauth_failure, %{})
+      |> get("/auth/google/callback")
+
+    assert get_flash(conn, :error) == "Sign in failed"
+
+    assert get_session(conn, :user) == nil
+    assert redirected_to(conn) == "/"
+  end
+
+  test "doesn't log user in if incorrect domain", %{conn: conn} do
+    # Domain is set to example.org in test config
+    auth = %{@ueberauth_auth | info: %{email: "test@bad.com", urls: %{website: "bad.com"}}}
+
+    conn =
+      conn
+      |> assign(:ueberauth_auth, auth)
+      |> get("/auth/google/callback")
+
+    assert get_flash(conn, :error) == "Sign in failed. Invalid domain"
+
+    assert get_session(conn, :user) == nil
+    assert redirected_to(conn) == "/"
+  end
+
+  test "Login page contains link to login with Google auth", %{conn: conn} do
+    conn =
+      conn
+      |> get("/auth/login")
+
+    assert html_response(conn, 200) =~ "Sign in with Google"
+  end
+
+  test "Logout page removes user from session" do
+    conn =
+      session_conn()
+      |> put_session(:user, %{email: "test@example.org", provider: "google"})
+      |> get("/auth/logout")
```

Comment:
Why does this test build its own `conn` instead of using the one from the test context?

Reply:
This is the `session_conn` function that we added to the `conn_case` helper, which builds a conn and adds the session into it. After some searching, this seems to be the way to be able to do a `put_session` before you've accessed an endpoint.

If you try a `put_session`, then it errors on that you need to call `fetch_session` first. If you put in `fetch_session`, it errors that the session isn't initialised, if you initialise the session, then everything works. So I've done that and added it all as a function in the ConnTest base helper (so that we can use it in all conn tests)